File: lestari/lestari/page/purchase_request_lis/purchase_request_lis.js
frappe.pages['purchase-request-lis'].on_page_load = function(wrapper) {
	new DevExtreme(wrapper)
}
DevExtreme = Class.extend({
	init: function(wrapper){
		var me = this
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: 'Purchase Order',
			single_column: true
		});
		this.list_po = []
		this.supplier = ""
		this.tujuan = ""
		this.pajak = ""
		this.ppn = ""
		this.no_faktur = ""
		this.submitted = ""
		this.combine = ""
		// this.page.add_inner_button('Update Posts', () => update_posts())
		// this.page.change_inner_button_type('Update Posts', null, 'primary');
		this.page.set_primary_action('Buat Purchase Order', () => this.submit(), { icon: 'add', size: 'sm'})
		this.page.set_secondary_action('Refresh', () => this.make(), { icon: 'refresh', size: 'sm'})
		this.page.add_inner_button('List Purchase Order', () => frappe.set_route(['List', 'Purchase Order']))
		this.page.add_field({"fieldtype": "Select", "fieldname": "tujuan","options": ['','Non Logam','Logam','Batu'],
			"label": __("Tujuan"), "reqd": 0,
			change: function() {
				me.tujuan = this.value;
			}
		}),
		this.page.add_field({"fieldtype": "Link", "fieldname": "supplier","options": "Supplier",
			"label": __("Supplier"), "reqd": 1,
			change: function() {
				me.supplier = this.value;
			}
		}),
		this.page.add_field({"fieldtype": "Check", "fieldname": "combine","default": 0,
			"label": __("Combine PO"), "reqd": 0,
			change: function() {
				me.combine = this.value;
			}
		}),
		this.page.add_field({"fieldtype": "Check", "fieldname": "pajak","default": 0,
			"label": __("Pajak"), "reqd": 0,
			change: function() {
				me.pajak = this.value;
			}
		}),
		this.page.add_field({"fieldtype": "Check", "fieldname": "ppn","default": 0,
			"label": __("PPn"), "reqd": 0,
			change: function() {
				me.ppn = this.value;
			}
		}),
		this.page.add_field({"fieldtype": "Data", "fieldname": "no_faktur","default": "",
			"label": __("No Faktur"), "reqd": 0,
			change: function() {
				me.no_faktur = this.value;
			}
		}),
		this.page.add_field({"fieldtype": "Check", "fieldname": "submitted","default": 0,
			"label": __("Submit"), "reqd": 0,
			change: function() {
				me.submitted = this.value;
			}
		}),
		$(".awesomplete>ul").css("z-index","3");
		// this.page.set_secondary_action(
		// 	__('Buat SPK Produksi'),
		// 	() => this.show_user_search_dialog(),
		// 	{ icon: 'add', size: 'sm'}
		// );
		this.make()
	},
	// make page
	make: async function(){
		let me = this
		DevExpress.localization.locale(navigator.language);
		let body = `<div class="dx-viewport">
			<div id="dataGrid"></div>
		</div>`;
		$(frappe.render_template(body, this)).appendTo(this.page.main)
		var employees =  await this.employees()
		// var formattedNumber = DevExpress.localization.formatNumber(employees.message., {
		// 	style: "currency",
		// 	currency: "",
		// 	useGrouping: true
		//   });
		// console.log(employees)		
		// DevExpress.localization.locale('id');
		$("#dataGrid").dxDataGrid({
			dataSource: employees.message,
        	keyExpr: 'idm',
			showBorders: true,
			
			allowColumnReordering: true,
			allowColumnResizing: true,
			columnAutoWidth: true,
			columnFixing: {
                enabled: true,
                 fixedPosition: "top"
            },
			scrolling: {
				columnRenderingMode: 'virtual',
				// mode: 'infinite'
			  },
			groupPanel: {
				visible: true,
			},
			 pager: {
                allowedPageSizes: [25, 50, 100],
                showPageSizeSelector: true,
                showNavigationButtons: true
            },
            paging: {
                pageSize: 25,
            },
			grouping:{
				autoExpandAll: true
			},
			selection: {
				mode: "multiple",
				allowSelectAll: true,
				selectAllMode: 'page' // or "multiple" | "none"
			}, 
			filterRow: { visible: true },
			headerFilter: {
				visible: true,
			  },
        	searchPanel: { visible: true }, 
			columnChooser: { enabled: true },
			export: {
				enabled: true
			},
			columns: [{
				dataField: 'no',
				width: 50,
				alignment: 'center',
				caption: 'No.',
				allowHeaderFiltering: false,
				sortOrder: 'asc',
			//   }],
			// ,{
			},{
				dataField: 'no_mr',
				format: 'string',
				alignment: 'left',
				sortOrder: 'asc',
				// width: 110,
				caption: 'No MR',
				groupIndex: 0
	   
			  }
			,{
				dataField: 'idmaterial_request',
				format: 'string',
				alignment: 'left',
				// width: 110,
				caption: 'ID MR'			   
			  }
			  ,{
				dataField: 'transaction_date',
				format: 'date',
				alignment: 'right',
				caption: 'Posting Date',
				// sortOrder: 'asc',
				// width: 110,
				
			  },
			//   {
				// dataField: 'schedule_date',
				// format: 'date',
				// alignment: 'right',
				// caption: 'Required Date',
				// sortOrder: 'asc',
				// width: 110,
				
			//   },
			//   {
			// 	dataField: 'status',
			// 	format: 'string',
			// 	alignment: 'left',
			// 	// width: 110,
			// 	caption: 'Status',
			// 	allowHeaderFiltering: false,			   
			//   },
			  {
				dataField: 'employee_name',
				format: 'string',
				alignment: 'left',
				// width: 110,
				caption: 'Employee Name',
				// allowHeaderFiltering: false,			   
			  },{
				dataField: 'department',
				format: 'string',
				alignment: 'left',
				// width: 110,
				caption: 'Department',
				// allowHeaderFiltering: false,			   
			  },			   
			  {
				dataField: 'item_code',
				format: 'string',
				// width: 150,
				caption: 'Item Code'
			  },
			  {
				dataField: 'description',
				format: 'string',
				// width: 150,
				caption: 'Description'
			  },
			  {
				dataField: 'qty',
				format: 'decimal',
				caption: 'Qty',
				allowHeaderFiltering: false,
			  },
			  {
				dataField: 'ordered_qty',
				format: 'decimal',
				caption: 'Ordered Qty',
				allowHeaderFiltering: false,
			  },{
				dataField: 'idm',
				format: 'string',
				alignment: 'left',
				// width: 110,
				caption: 'IDM',
				// groupIndex: 0
	   
			  }
			  ],
			  onSelectionChanged(e) {
				e.component.refresh(true);
				if(e.currentSelectedRowKeys[0] != null){	
					me.list_po.push(e.currentSelectedRowKeys[0])
				}
				if(e.currentDeselectedRowKeys[0] != null){
					me.list_po = me.list_po.filter(data => data != e.currentDeselectedRowKeys[0])
				}
				// console.log(e)
				// console.log(me.list_spk)
			  },
			//   summary: {
			// 	totalItems: [{
			// 	  name: 'SelectedRowsSummary',
			// 	  showInColumn: 'qty',
			// 	  displayFormat: 'Qty: {0}',
			// 	  summaryType: 'custom',
			// 	},
			// 	],
			// 	calculateCustomSummary(options) {
					
			// 	  if (options.name === 'SelectedRowsSummary') {
			// 		if (options.summaryProcess === 'start') {
			// 		  options.totalValue = 0;
			// 		}
			// 		if (options.summaryProcess === 'calculate') {
			// 		  if (options.component.isRowSelected(options.value.name)) {
			// 			options.totalValue += options.value.qty;
			// 		  }
			// 		}
			// 	  }
			// 	},
			//   },
			  onExporting(e) {
				const workbook = new ExcelJS.Workbook();
				const worksheet = workbook.addWorksheet('Employees');
		  
				DevExpress.excelExporter.exportDataGrid({
				  component: e.component,
				  worksheet,
				  autoFilterEnabled: true,
				}).then(() => {
				  workbook.xlsx.writeBuffer().then((buffer) => {
					saveAs(new Blob([buffer], { type: 'application/octet-stream' }), 'TransferSalesman.xlsx');
				  });
				});
				e.cancel = true;
			},
			// onSelectionChanged(selectedItems) {
            //     const data = selectedItems.selectedRowsData;
            //     if (data.length > 0) {
            //         // $('#selected-items-container').val(
            //         // data
            //         //     .map((value) => `${value.linkPopup}`)
            //         //     .join(', '),
            //         // );
			// 		console.log(data)
            //     } else {
            //         $('#selected-items-container').val('NULL');
                    
            //     }
       
            // },
		});
		
	},
	employees: function(){
		var data = frappe.call({
			method: 'lestari.lestari.page.purchase_request_lis.purchase_request_lis.contoh_report',
		});

		return data
	},
	submit: function(){
		var me = this
		// console.log("test"+me.supplier) // disini keluar test doang masih kosong
		frappe.call({
			method: 'lestari.lestari.page.purchase_request_lis.purchase_request_lis.make_po',
			args: {
				'data': this.list_po,
				'supplier': me.supplier, // disini kosong why???
				"tujuan":me.tujuan,
				"pajak":me.pajak,
				"ppn":me.ppn,
				"no_faktur":me.no_faktur,
				"submitted":me.submitted,
				"combine":me.combine
			},
			callback: function(){
				me.make()
			}
		})
	}
})

File: lestari/lestari/page/purchase_request_lis/purchase_request_lis.py
# Copyright (c) 2021, Patrick Stuhlmüller and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe.utils import *
import json

@frappe.whitelist()
def make_po(data,supplier = None, combine = None, pajak = None, ppn = None, tujuan = None, no_faktur = None, submitted = None):
    list_name = tuple(json.loads(data))
    # frappe.msgprint(str(list_name))
    mr_list = frappe.db.sql("""
        SELECT 
        a.transaction_date, a.name, a.schedule_date, 
        b.name as IDM, b.item_code, b.schedule_date, b.item_name, b.description, b.keterangan, b.qty, b.uom, b.stock_uom, b.conversion_factor
        FROM `tabMaterial Request` a JOIN `tabMaterial Request Item` b ON a.name = b.parent WHERE b.name IN {}
    """.format(str(list_name)),as_dict=1)
    frappe.msgprint(str(mr_list))
    new_doc = frappe.new_doc("Purchase Order")
    new_doc.naming_series = "PO.YY.MM.DD.###"
    new_doc.transaction_date = mr_list[0].transaction_date
    required_date = mr_list[0].schedule_date
    new_doc.tujuan = tujuan
    new_doc.pajak = pajak
    new_doc.ppn = ppn
    new_doc.no_faktur = no_faktur
    new_doc.supplier = supplier
    for row in mr_list:
        baris_baru = {
            "item_code": row['item_code'],
            "schedule_date": required_date,
            "item_name": row['item_name'],
            "description": row['description'],
            "keterangan": row['keterangan'],
            "qty": row['qty'],
            "uom": row['uom'],
            "stock_uom": row['stock_uom'],
            "conversion_factor": row['conversion_factor']
        }
        new_doc.append("items",baris_baru)
    new_doc.flags.ignore_permissions = True
    new_doc.save()
    if submitted and submitted == 1:
        new_doc.submit()      
    frappe.msgprint("PO Berhasil dibuat dengan Nomor"+str(new_doc))

@frappe.whitelist()
def make_purchase_order(source_name, target_doc=None, args=None):
	doclist = get_mapped_doc(
		"Material Request",
		source_name,
		{
			"Material Request": {
				"doctype": "Purchase Order",
				"validation": {"docstatus": ["=", 1], "material_request_type": ["=", "Purchase"]},
			},
			"Material Request Item": {
				"doctype": "Purchase Order Item",
				"field_map": [
					["name", "material_request_item"],
					["parent", "material_request"],
					["uom", "stock_uom"],
					["uom", "uom"],
				],
			},
		},
	)

	return doclist

@frappe.whitelist()
def contoh_report():
    po = []
    # list_doc = frappe.get_list("Form Order", filters={'docstatus':1},order_by="posting_date DESC",limit = 5000)
    list_doc = frappe.db.sql("""
        SELECT a.name as ID, a.idmaterial_request, a.department, a.employee_name, a.status, a.transaction_date, a.schedule_date,b.name as IDM, b.ordered_qty, b.item_code, b.`description`, b.qty
        FROM `tabMaterial Request` a
        JOIN `tabMaterial Request Item` b 
        ON a.name = b.parent
        WHERE a.docstatus = 1 AND a.status = "Pending" AND a.material_request_type IN ('Purchase','Partially Ordered')
        
    """,as_dict = 1)
    # frappe.msgprint(str(list_doc))
    no = 0
    for row in list_doc:
        no+=1
        baris_baris = {
            'no' : no,
                'no_mr' : row.ID,
                'idm' : row.IDM,
                'idmaterial_request': row.idmaterial_request,
                'status' : row.status,
                'employee_name' : row.employee_name,
                'department' : row.department,
                'transaction_date' : frappe.format(row.transaction_date,{'fieldtype':'Date'}),
                'schedule_date' : frappe.format(row.schedule_date,{'fieldtype':'Date'}),
                'item_code' : row.item_code,
                'description' : row.description,
                'qty' : row.qty,
                'ordered_qty' : row.ordered_qty
        }
        po.append(baris_baris)
    # frappe.msgprint(str(po))  
    return po   